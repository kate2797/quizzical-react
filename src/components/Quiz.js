/* eslint-disable react-hooks/exhaustive-deps */
import React from "react";
import Question from "../Question";

export default function Quiz() {
  const [questionData, setQuestionData] = React.useState([]);
  const [score, setScore] = React.useState(0); // number of answers they got correct

  // this component will display the score etc.

  /**
   * gets data from the API and transforms them into an array of objects
   */
  React.useEffect(() => {
    fetch("https://opentdb.com/api.php?amount=5&type=multiple")
      .then((res) => res.json())
      .then((res) =>
        setQuestionData(() => {
          let index = 0;
          return res.results.map((data) => {
            return {
              id: index++,
              question: data.question,
              answers: [...data.incorrect_answers, data.correct_answer].sort(
                () => Math.random() - 0.5 // shuffled answers
              ),
              correctAnswer: data.correct_answer,
              userAnswer: "",
              isRevealed: false, // needed for styling
            };
          });
        })
      );
  }, []); // I want to make the API call just once

  function selectAnswer(questionId, answer) {
    // get that question, update its property

    console.log(questionId);
    console.log(answer);

    setQuestionData((prevData) => {
      return prevData.map((question) => {
        if (question.id === questionId) {
          return {
            ...question,
            userAnswer: answer,
          };
        } else {
          return question; // must specify the unmodified version too
        }
      });
    });
  }

  console.log(questionData);

  const questionElements = questionData.map((question) => {
    return (
      <Question
        key={question.id}
        id={question.id}
        question={question.question}
        answers={question.answers}
        correctAnswer={question.correctAnswer}
        userAnswer={question.userAnswer}
        isRevealed={question.isRevealed}
        selectAnswer={selectAnswer}
      />
    );
  });

  return (
    <>
      <div className='quiz'>
        <div>{questionElements}</div>
      </div>
      <button className='button-main'>Check Answers</button>
    </>
  );
}
