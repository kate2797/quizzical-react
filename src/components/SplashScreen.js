import React from "react";

export default function SplashScreen(props) {
  return (
    <section className='splashscreen'>
      <div className='splashscreen-container'>
        <h1>Quizzical</h1>
        <p>Can you get all of them right?</p>
        <button onClick={props.startQuiz}>Start Quiz</button>
      </div>
    </section>
  );
}
