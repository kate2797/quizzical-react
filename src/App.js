import React from "react";
import SplashScreen from "./components/SplashScreen";
import Quiz from "./components/Quiz";

/**
 * Starting the quiz pulls questions from the API
 * Conditionally render the next section
 */

export default function App() {
  const [hasStarted, setHasStarted] = React.useState(false); // state to handle SplashScreen

  /**
   * flips the state, so that we now render the correct component (Quiz)
   */
  function startQuiz() {
    setHasStarted(true);
  }

  return (
    <main>
      {!hasStarted && <SplashScreen startQuiz={startQuiz} />}
      {hasStarted && <Quiz />}
    </main>
  );
}
