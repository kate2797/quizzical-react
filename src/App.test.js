import React from "react";
import { mount, shallow } from "enzyme";

import App from "./App";
import SplashScreen from "./components/SplashScreen";
import Quiz from "./components/Quiz";
import toJson from "enzyme-to-json";

// setup
const setHasStarted = true;

/**
 * test suite 1 - rendering (regression tests)
 */
describe("test rendering of components", () => {
  it("renders App component without crashing", () => {
    shallow(<App />);
  });

  it("renders SplashScreen component without crashing", () => {
    shallow(<SplashScreen />);
  });

  it("renders Quiz component without crashing", () => {
    shallow(<Quiz />);
  });
});

/**
 * test suite 2 - passing props (unit tests: individual componets or pure functions)
 */
describe("test passing props", () => {
  const quizWrapper = mount(<Quiz setHasStarted={setHasStarted} />);

  it("accepts function props", () => {
    expect(quizWrapper.props().setHasStarted).toEqual(setHasStarted);
  });
});

/**
 * test suite 3 - logic (unit tests)
 */
// left out - no complicated logic lives in App

/**
 * test suite 4 - snapshots (regression tests)
 */
describe("test snapshots", () => {
  it("App snapshot", () => {
    const tree = shallow(<App />);
    expect(toJson(tree)).toMatchSnapshot();
  });
});
