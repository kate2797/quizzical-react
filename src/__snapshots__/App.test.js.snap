// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test snapshots App snapshot 1`] = `
<Fragment>
  <main>
    <SplashScreen
      startQuiz={[Function]}
    />
  </main>
</Fragment>
`;
